src/common/cm/cm_trace.cpp

```diff
@@ -73,15 +73,17 @@ TransposeMatrix
 */
 void TransposeMatrix( const vec3_t matrix[ 3 ], vec3_t transpose[ 3 ] )
 {
-	int i, j;
+	transpose[ 0 ][ 0 ] = matrix[ 0 ][ 0 ];
```

Comment:
Hmm, yes, this one may be stupid.

Both gcc and clang know how to convert this code into SSE code with both implementation (loops and unrolled), while MSVC fails to convert this code into SSE with both implementation (loops and unrolled), [see](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahUAUgBMAIUtXyKxqiIEh1ZpgDC6egFc2TA3cAGQImbAA5PwAjbFIpHnIAB3QlYhcmL19/A2TU5yEQsMi2GLieBIdsJ3SRIhZSIky/AIrsR3ymWvqiQojo2Pj7Ooam7IrhntC%2BkoHygEp7dB9SVE4uH1STAGoANyqAfSItswBmABEtmnp0FiJTmy0AQQ3Q4F2qk8Pj8/fUQ7MAKxWE6As73MxPCGPHboAiYLYAFVIxiUuWwAFlbqQCOoIFsMEwlEc9qhPkc2FicYCrFsTscAWdyL8yVsiMjCWjqbT6Rc5lCzAB2B6PCEATlCRwITN04MhctFglIWzxBG%2BFy09y2qtOHlpmoItlsWz58sFwrFYsVyq2ujVWw1Jxptp1esdNsN1mN/K0ootQu9vp9YrZKM5QK1PK5zoZdopbKp4ejYPD2oZspFQaDArBpuz/LzcuhsPhSNDKQxlPUFjxBKJzK%2Bcex6i5dNBTJJLJDHPLLZ5Xsh/vlXdRPfDWkjY77pwujYTNPHoK5C7TjotQ/ZI5US4nNJ4U5%2Bs%2Bb4b3i8ni9X67LW7PafDFn3M8rXPvp/nkYvGYzouHYd3O/tD5bIe26vlsJ4ruaPo/qOf6geBFzTkBT7Hv%2B8HpsGG6/mB/4vjGiHAXeqHvpBX7QdeNK4cmb6grGyHUbeFHEQGZHYM%2BRE0fhdHYaBlHoVBmEwVsvGERxB5ccJjHnua2ZcAs9DcAC/ABFwOjkOg3AeB6NJKEsKzYMcFgnHw5BENoskLAA1iA0haAAdAAbCcAAc9kAicJxaBY0iihYoqyPJXDSPwbAgPZFi2QC5DKap6lcPwSggFoJlmQscCwCgSxEIkPhEJQ1AYGwiQMLE4TsGs6guQAtPZ0hbMAqCoGB4UWPw2CECQ2KYAYgjCGIEicDIcjCMoaiaCpegGAA7siiTcHwckKUpZlqdwADyOXZUc6A0FsFX2dVtX1Y1PDNcqXiFcVSqWEZcz8KZ41zAsSDYCwOBxBAC2BcFIAnAKEVRctsXxYlyUPZZP1ObZTknSchlGZ5Wgef53AnEt40rXFoM6KliBpcgaDoBdjAUFQEAFUVxMgMAJ0WOQOA7AQqwAGoENgk2rYkzBzfwdD0EQsQJRAUTLVEoT1AAntz5Ci6wpDi6tUS6FU93GQVHDCKtTD0JL6M4BSJiSLrBCkMrBB7Al6PYOoVQ5WsxkSm0y30AQUTInLXg4Mt8YhbwskCEYwBKCzbMc1zvtDaI4iSINPWKCoGjLfoCRGCYaBaYYLsJZACzoIkHQW5Vq2o/YbSm64EDuKMLTkMEUzFKUOQpGkQhV43eTpL09ezCX7Q1BMretL3QhdA0nf9GUQzdAPk%2Bj3X49SAsOnLKsC%2BGItAPo7Fu1VTVdUNU1tn3hA%2BDEGQBk3XdKXgxYf1OQC0gnA/JwAjVjmOQKa9feQIUAkl0X8EDewIN7rY3IHjCAKByaXTymTQmFMBiM1QBYTyPABQJHpozbAwd2ac2UsZXm/NSCC2FujGWEspZkLlgrJWTgpZq2YEQTW2tlp62MMAQ2qlCAm2qObZaVsbb8ylg7AKqlnauwlh7NYqlvbcwWFcFggdsGhzwTzeQfVo6yFjiNBO6N9D%2BTYWnawtgM5RCzh9NSed0gF1Wi1HuZcAgVyYJ4bwzRAhOLHjMCeuRm4ZBcWMJITcOgeIboPexI9Gh%2BOrpUaow8JjBO7kSKekSDCJNnkUeePBF66RXpkz%2BikN4xW4Nvfau9EFCRQbZP6e5j7tTPtdTJl8wbkGeq9AY5iApBW/iAX%2BBSAHcGBklEB5lyBWRvlDe%2Bj93Iv2kG/X6n9i7/wxo00B4DcaQLgdA0mUDKaoBTtTeyWgkoYOZqzHBYd8EMEIcQkWYs5YUNufLRWys6GE3VowrWOtOHYH1uwqR/AuGm14Zba2qBbZCOEI7dGYi3bi0kV7bEPt5r%2BwUUHU5yipax3UQNTR8htFjVUvoE4hgDHmCMdYExZic6WKENY2x0SOhuCcdPWu6TPFtx8dPbxQS55stCTEzo/dkl8o6OE%2BJE9UkRKyFEuJPKQlZOXgNT6%2BTFlbz2gdfEeywL2VsnZccNTT5XUMg0rGwyWlvWoJ9TpP8/6A36UAwZKUwFrIJkTWIMDtkDB2KgRIiR9g7B4KKfY7l9jqBqnTbADMTkh1wRiy5AtqAkNUpQnWxlk3UOeeHehGsPksO%2BWwjh/zjaAuwBbVS/DQWCPDsIp2LsYVwvRjI32ciA6oujec1RvUo7YojnixOIBaZ7MMTYclYjKUWPziAAA9DYux/LGXOKlW4zAYr2UdE5YEjusqEml35eE6e9K%2B7dBXeMJJi6T1pOmHKxYCrV4BWVbargu1d5sCUF63YAbbLuWVCfDq59jVDMeiMn6JxP3uTA%2BB8Dn8rXdJtZvO1CUHVNKsu5UDEG0NEoCgsh9yzhkBVsSqu1AGFh7CIekayQA)  `¯\_(ツ)_/¯`

Damn…

We would probably benefit from glm dedicated function instead, that would probably do the related parallelized code on purpose whatever the compiler.

Other unrolling looks to benefit a lot though (for example the ones like “_offset for capsule center_”, [see](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahUAUgBMAIUtXyKxqiIEh1ZpgDC6egFc2TKQtydwAZAiZsADk/ACNsUikAdnIAB3QlYhcmL19/QNT0zKEwiOi2OISeZIdsJyyRIhZSIhy/AJ4gmrqhBqaiEqjY%2BKT7RubWvI7RvoGyiqSASnt0H1JUTi4fDJMAagA3WoB9Ih2zAGYAER2aenQWInObLQBBLfDgfdqz49PLz9RjmYAKxWM7Ai6PMwvKHPPboAiYHakbz0CD/b4nNjhJTkdE/NgsdQ4vEndA0GgqIi4g6oDE7ABUGQAXtgdgtTokns8oQBOAD0fJ2ZIp2BOglIO3EKSUPkYkuYRHiMN54p2aPCJwIvyuWkeOy15w8OzBZys%2BtstjZyq0PLMnOttptwspwLNBqBV3OVzRWKYSld%2BtOHtO1h2BKJAfdV3Z9J2WgAdEDIdCbbznaKA1og56/j7sZnsyGzeH/SC44WY3HE8nuambczsAXwZHC16w/my1GdgBaIXkl2d7M13lp/sZss8Vu59t%2BgOT8FFsOE0tm%2BfBysJpOmh28htzoeDhdtksthe99MPQ8e4d1nkXgMWKfemcrnaPo%2Bhk9l9/rhlVrdciOiQQtCwHWrC8KIj4TDIvQqIki%2B1JfPiy5IbSPwXmhdKMgQLJshygE2gKfYimKZCSiw0qyqy6zCEqLw7IxBE7k6Y6XmaWZHtOvqvpxwaWl%2BHEVn%2Bm43o6d5sfuXHPjxUn8Z%2By5ydGInVtuKbife35PmqL4Pq2CkRlpC4bqphG1uJe5lnxEJWdpsm2WeJEDkJzZqeZu64Y2DnXhOdkdquhbnpJvmuYRFmeU2Plmj%2BOZXPZ0WBU544JaF4EjvWEUhVF5bSUuhkucGQWkZFEJuelPKWQFzZZbFeWvmuVxFc5OwNWJHkskpem5YJb6JZpKXXm5TEciB3LAVwSz0NwQL8AEXA6OQ6DcB4FqhkoKxrKylhnHw5BENoE1LAA1iA0gJgAbGcAAc51AmcZxaBY0g8hYPKyFNXDSPwbAgOdFiJuQc0LUtXD8EoIBaHtB1LHAsAoCsRApD4VJUBAGBsCkDDxJE7AbOoN3dud0g7MAqCoC1/0WPw2CECQpAIgYgjCGIEicDIcjCMoaiaPNegGAA7qQlHcHwk3TbNB2LdwADyyNI6SNA7Pj52E8TpPkx08aPhAXgY1jErbTwCz8PtvMLEsSDYCwOAJBAYufd9IBnIkANA/wINgxDUNm8dTtXfGV0dGcFj3Twj1aA973cGcEu81LoPezoMOILDyBoOgeuMBQqPo5jWcgMAHRBDgewEOsABqBDYPz0spMwIv8HQ9CKqQ4MQDEksxOETQAJ4N%2BQXesKQPfSzEui1Kbu3oxwwjS0w9B93HOAEiYkhLwQpATwQBzg3H2DqLUyMbLtGrYB9C30AQMRC8PXg4JLRD0z9vATQIRjAEolfV7X9cvxzojiEkOzJmigVAaElvoHghhjCmFWnYS%2BMRwaQCWOgFIzghC727NLGO9gz5b1cBAdwEx2jBCYJgWYQxKgFAyOg7I3g2gGDSDQrIFDyjDCgV0WhvRxj0MmLgxwXCxj9HCIMNhVClBCOIQYCRMwRFzHYUsdaqx1hSHtjNQGksQZKwJkTEmZMKZazVPgYg5FDbG0TodcgJ0LAuyukCaQZwHFnCBETS6l1kgfS%2BuQH6QJIZu3jp7SGpsk7kFThAFAud9aUGoJE/OZdUAWEelUKBJcy7YC/jXOuc1dpNxbm3DucdB6937kU4eo9x5OH7tPBUc8F6S2XjAteC1CCbzqDvSW%2B9D6Kn7qfc%2B/AEE3x7nfDYC1H4EGfqLN%2BLAP4ZJ/tkxu8gWZANkCArm4C476HejAtAcDDBXyQXbRaaCsiYOllTfh%2BCAiELIVIqBoQ5GUMYYUWhtzqFFCYKw%2BYHC8HdCYNwlovCSGcPqEIz57Dpg8NyEC0FDyxGqOWMotmajY7A24NolWuj4lviSfGF2k4IDGLpiGHa5jgmWMttbYYhzPGO18RouOHt7BezJebKxIAbEB3sY4%2B6LjpBuOdoYaOKL3bcBNtDUJKcU4RIznneI0S0Yyv1mgIwJgeDnS0JDVJFcq6ZN/jkhgeTqAFIWqUxeu1TXlInlUjOM8iC1MXs07AK9gBNP4C0re7S94H1QEfHpdE%2Bl7Ovr3YZD8n4NyWDcaZn8dVzP7iApZbMVnyDWTzBa%2BgzjQJMDs6wthA0HJQccjB3AsHnOBQQohgKDD3NKI8qBTD3mvPrbQsF4ifmCL6K8stfyYU1rhRwyRlb%2B2yN7V8xRG0VFG0FVwdR/itHK1VpKFVHw1XxgTFmAltNTEhyNmKn25AKU22oPbLxPi/GaNFUyoJ4qwnSsznKnOir857FQCkFIhw9g8B5Ice6hx1BE3IFq9JMaslxoNfEfJndu7DxKVBkeY8rV/2qbPeeDq3VOsaSMt1G8PXYF3gtTpPrul/16ZLAZwasCYb2mGl%2BEb37Ru/iBv%2B8bAGJv/imiB7LM2wJzdYPN8AC20PBnyM5FzfluBuYO0h5DYVfLeS8yTTaWEyfBV2/5na20guHaI2TMjIUMKHc0Ft8KlGbXhR9Gd56uBK10WwJQz79ifvjPdIxm6DbbtJdDX290nP3V835vzU6T0gDpbOi94Mr17pOt5/zMXfNTpwaFhOLL7bnMS7upOSwDityyKdIAA%3D%3D).

Reply:
Why would you even want to use SSE here? There are no floating point operations; it's just copying memory.

Reply:
> Other unrolling looks to benefit a lot though (for example the ones like “_offset for capsule center_”, [see](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahUAUgBMAIUtXyKxqiIEh1ZpgDC6egFc2TKQtydwAZAiZsADk/ACNsUikAdnIAB3QlYhcmL19/QNT0zKEwiOi2OISeZIdsJyyRIhZSIhy/AJ4gmrqhBqaiEqjY%2BKT7RubWvI7RvoGyiqSASnt0H1JUTi4fDJMAagA3WoB9Ih2zAGYAER2aenQWInObLQBBLfDgfdqz49PLz9RjmYAKxWM7Ai6PMwvKHPPboAiYHakbz0CD/b4nNjhJTkdE/NgsdQ4vEndA0GgqIi4g6oDE7ABUGQAXtgdgtTokns8oQBOAD0fJ2ZIp2BOglIO3EKSUPkYkuYRHiMN54p2aPCJwIvyuWkeOy15w8OzBZys%2BtstjZyq0PLMnOttptwspwLNBqBV3OVzRWKYSld%2BtOHtO1h2BKJAfdV3Z9J2WgAdEDIdCbbznaKA1og56/j7sZnsyGzeH/SC44WY3HE8nuambczsAXwZHC16w/my1GdgBaIXkl2d7M13lp/sZss8Vu59t%2BgOT8FFsOE0tm%2BfBysJpOmh28htzoeDhdtksthe99MPQ8e4d1nkXgMWKfemcrnaPo%2Bhk9l9/rhlVrdciOiQQtCwHWrC8KIj4TDIvQqIki%2B1JfPiy5IbSPwXmhdKMgQLJshygE2gKfYimKZCSiw0qyqy6zCEqLw7IxBE7k6Y6XmaWZHtOvqvpxwaWl%2BHEVn%2Bm43o6d5sfuXHPjxUn8Z%2By5ydGInVtuKbife35PmqL4Pq2CkRlpC4bqphG1uJe5lnxEJWdpsm2WeJEDkJzZqeZu64Y2DnXhOdkdquhbnpJvmuYRFmeU2Plmj%2BOZXPZ0WBU544JaF4EjvWEUhVF5bSUuhkucGQWkZFEJuelPKWQFzZZbFeWvmuVxFc5OwNWJHkskpem5YJb6JZpKXXm5TEciB3LAVwSz0NwQL8AEXA6OQ6DcB4FqhkoKxrKylhnHw5BENoE1LAA1iA0gJgAbGcAAc51AmcZxaBY0g8hYPKyFNXDSPwbAgOdFiJuQc0LUtXD8EoIBaHtB1LHAsAoCsRApD4VJUBAGBsCkDDxJE7AbOoN3dud0g7MAqCoC1/0WPw2CECQpAIgYgjCGIEicDIcjCMoaiaPNegGAA7qQlHcHwk3TbNB2LdwADyyNI6SNA7Pj52E8TpPkx08aPhAXgY1jErbTwCz8PtvMLEsSDYCwOAJBAYufd9IBnIkANA/wINgxDUNm8dTtXfGV0dGcFj3Twj1aA973cGcEu81LoPezoMOILDyBoOgeuMBQqPo5jWcgMAHRBDgewEOsABqBDYPz0spMwIv8HQ9CKqQ4MQDEksxOETQAJ4N%2BQXesKQPfSzEui1Kbu3oxwwjS0w9B93HOAEiYkhLwQpATwQBzg3H2DqLUyMbLtGrYB9C30AQMRC8PXg4JLRD0z9vATQIRjAEolfV7X9cvxzojiEkOzJmigVAaElvoHghhjCmFWnYS%2BMRwaQCWOgFIzghC727NLGO9gz5b1cBAdwEx2jBCYJgWYQxKgFAyOg7I3g2gGDSDQrIFDyjDCgV0WhvRxj0MmLgxwXCxj9HCIMNhVClBCOIQYCRMwRFzHYUsdaqx1hSHtjNQGksQZKwJkTEmZMKZazVPgYg5FDbG0TodcgJ0LAuyukCaQZwHFnCBETS6l1kgfS%2BuQH6QJIZu3jp7SGpsk7kFThAFAud9aUGoJE/OZdUAWEelUKBJcy7YC/jXOuc1dpNxbm3DucdB6937kU4eo9x5OH7tPBUc8F6S2XjAteC1CCbzqDvSW%2B9D6Kn7qfc%2B/AEE3x7nfDYC1H4EGfqLN%2BLAP4ZJ/tkxu8gWZANkCArm4C476HejAtAcDDBXyQXbRaaCsiYOllTfh%2BCAiELIVIqBoQ5GUMYYUWhtzqFFCYKw%2BYHC8HdCYNwlovCSGcPqEIz57Dpg8NyEC0FDyxGqOWMotmajY7A24NolWuj4lviSfGF2k4IDGLpiGHa5jgmWMttbYYhzPGO18RouOHt7BezJebKxIAbEB3sY4%2B6LjpBuOdoYaOKL3bcBNtDUJKcU4RIznneI0S0Yyv1mgIwJgeDnS0JDVJFcq6ZN/jkhgeTqAFIWqUxeu1TXlInlUjOM8iC1MXs07AK9gBNP4C0re7S94H1QEfHpdE%2Bl7Ovr3YZD8n4NyWDcaZn8dVzP7iApZbMVnyDWTzBa%2BgzjQJMDs6wthA0HJQccjB3AsHnOBQQohgKDD3NKI8qBTD3mvPrbQsF4ifmCL6K8stfyYU1rhRwyRlb%2B2yN7V8xRG0VFG0FVwdR/itHK1VpKFVHw1XxgTFmAltNTEhyNmKn25AKU22oPbLxPi/GaNFUyoJ4qwnSsznKnOir857FQCkFIhw9g8B5Ice6hx1BE3IFq9JMaslxoNfEfJndu7DxKVBkeY8rV/2qbPeeDq3VOsaSMt1G8PXYF3gtTpPrul/16ZLAZwasCYb2mGl%2BEb37Ru/iBv%2B8bAGJv/imiB7LM2wJzdYPN8AC20PBnyM5FzfluBuYO0h5DYVfLeS8yTTaWEyfBV2/5na20guHaI2TMjIUMKHc0Ft8KlGbXhR9Gd56uBK10WwJQz79ifvjPdIxm6DbbtJdDX290nP3V835vzU6T0gDpbOi94Mr17pOt5/zMXfNTpwaFhOLL7bnMS7upOSwDityyKdIAA%3D%3D).

The two variants look quite similar to each other in gcc and clang. Can you explain how it is benefiting a lot?

Reply:
Hmm, I made a copy paste mistake in the tested code in the previous link. Anyway the rolled implementation is still less good.

Clang produces the same optimal code on O2 and O3 either with rolled or unrolled code.

GCC only produces optimal code with unrolled code, on O2 the rolled code just keeps the loop, O3 does a very weird thing, doing a large compute to select among multiple alternate codes with possible recursion…

MSVC uses a bit more instructions on unrolled loop, that's weird, both with O2 and Ox. Anyway it's not that bad, less bad than what gcc does, and we don't distribute any release built with MSVC anyway.

The unrolled code is not only better on gcc while being as good on clang, it's also more predictable.

For O2 test, [see](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahUAUgBMAIUtXyKxqiIEh1ZpgDC6egFc2TA3cAGQImbAA5PwAjbFIQAFZyAAd0JWIXJi9ffwMUtOchELDIthi4xIdsJwyRIhZSIiy/AJ57bEcCplr6oiKI6NiE%2BzqGppzWpRHe0P7SwfiASnt0H1JUTi4fNJMAagA3KoB9Ih2zAGYAER2aenQWInObLQBBLdDgfaqz49PLz9RjmZ4lYzkCLo8zC9Ic89ugCJgdqRvPQIP9vic2KElOQ0T82Cx1NjcSd0DQaCoiDiDqh0TsAFRpABe2B2C1OAHYns9IQBOQSkHao0InAi/K5aR47UXnDw7UFnKxS2y2VnQ3lmTlqrQ83mk8nYB7AqWneJXc5XVGYphKIGK6Wm07WHb4wm241g1n0nZaAB08QhUO1vKZ2DdWhN4KN9rNfytNqjEZ2AFodnqKW7owHuUHtSG3TwIxnE%2BbnQT43bEym0wai2Cs%2Br2eCoY2uEt6Nx4vwAlwdOR0NwPMqnUoVmsWZYznxyERtK2lgBrKQADh9WgAbGv2Vp4jyl%2Byl0v4vEeGdDNxpPw2AktORu73%2B1x%2BEoQDeZz3W%2BQ4LAUBgcPhiGQlDUHQjCsBw3BToIwhiBInAyHIwjKGomjvuQ%2BgWIYximEOdiVNUrgQO4YwtOQwQzCUZRDHk6RCMRuSpDRTB9BR8xtB0NRTHREztFUnTdA0zEDOUww9FxIkCeRQkJEsI6rOs3BvLs1I/CWNx3A8CrQopHzUrSJbKYaIJ1ppgYwnCCI%2BEwSL0CixLOliVJfHiZaOTSPzVpSdkMgQzKehqXINgFOY8h5YbFn8lpYmFHoqi65beombJ0t6fr1sFoVGgWMURfZ1r5sWTpxflHpJSl/omdmOrahlioWOFFq5fFdUxYVZZus1Dqlb65VBZVwY%2BaGRrhnWQ31Y10UOlWZLpqNxlBVVPJ5rNprFQ6JZxqtVxTfqhk7FlK0VequYDRNka1WNG1Gh1W2ptNNZXYWh2BgtS2KvtZ0JdlVxFctN01Z9B3zf1zKbZtYqlq6mWVrdO2g09L0nVDI3nV9ENNdD/3XVmOw4xyTbci2bYdl2s59gOOE7LJY6OmcFj8G%2BOgLEsSDYCwOBxBARNcBe5BXvEN53vwD5Pi%2B06zkz5CLjwPI%2BuyPBaEu27smcPIWPL8SbmeXBnCTqHC2L75LF%2ByBoOgbBJAwsRARAGDm5bcTADwFgYTgewEOsABqBDYAA7gA8kkzAQfwIFELEz4QFEpNRKE9QAJ7B%2BQMesKQcd%2B1Eui8YntscMIftMPQCeoTg%2BImJIxcEKQvEEAcz6odg6hVD4YeJ8K7Sk/QBBRKQ8deDgpNEKQBBXrwH43CwwBKF7vsB0Ho8IaI4iSPBUGKCoGik/orRGCYaA4YYXfPpASzoEknR10mft02x1cEUR3jNIETCYIJcxxK01GdGJn8ZK/lHcexIQ/FGgP3GDffCXQph/0GBMTioCSKTB6NA9%2BMlRzyR4FzTst5SYPh2OoJca4kxrmkDsYAqBUB7QsD6OqEB/wkAFBODB9NxYLikFQs4a54jsmkBYHhKseEWDOOyLWPMrzywFjg7gItXzi0/IgCAP4zYW0YBQKgNslH2xAI7Z25BXbu2wNPf2gduyQQYGHUgEco6oWTvHRONjU7p0zk4bOZtc5EHzoXUmJcsLl17IQKu1Ra6kwbk3Fu8827tlQp3buvcsAbF7IPYewcljj0noY2eJiQ7yBgsvWQq8kIb1QvoYRO9sLWFsAfKIR9OZ9jPhkC%2BV8nw8QgW4Z%2BYkyLFCkh/BiX94H0XyL/SSb8DB4T4nA7ICDmljKQUM/%2B4kQETJGVA2ZMDUFyU4BYTBut7zcDwQQohJCyEUKdtQwUdCyA002cww2zNWbs2oKwk5KsLD8y0NIfma4tAWE%2BWuERl4pBaAkXrKR9hRYMw/MbSFii7YqOtrbZRgxUClJ4J8m8ejPbeyMXPUx9BzGWOjrHVOdjCVpwzlneeOdmDuILkXPx2BS7AF8fwfx1cgn10bqgZuGwpwRI7l3Huqc%2B7xPpkPEefAUlGDSZijJidV45Lgnk%2BQBSUK9n0KeUpe9ynWEqdUk%2BdShANOvqMjIrTPB9NaB02Ycyf60XNckHpgzOnDIAbfSBok7XGqAcsp1czEGjA9d6q1qzljrKkFs7BwKuB7MIcQnYSKsJ7TXKuVcZzCD0Jpkwg2jNbls0GDUxcLy/RrheWcJc0geTK03DyLQ6rzz/PERGnZj5QUyMNnI%2BACjTYwqtmo%2BFmj40mBRYC3R2A3YYpnsY2VZjw7UCsb2exRcpwLsceSqclK840q8fSnxwrdGV1ZdgOuvYQmcrCTy4Q7con8tif3VCiSxVj0lVPaVk755yqXgqheyrN5DA1eYLVdhom6tqefbgl8jVTJNYRNpdrLUsXfvagZtrFndKQ0xFZCHPVuv9Sh8B0yJI%2BpgfMsSfrpiEZQSGscIBNlaywYLMmUb8ExpIQOj4KLk3hloWmi5E4rlZrnOQFmuaOasMLRrGQ8RnYK0kzyI8fzeYAqBU26R/GjbyOhQi1R1A%2B0qJAHsVASQkiHD2NLQ4ZxvjqGISOsdBiX3YpDtOixs6CUp0XfwZdZLnEUtcVSjxtLmXbrLrullgTD3BI5Vy1uF7Im9miQKuOQqB6iuSQIJ96TX2QWyR%2BqQirELrxVXoajmFd7/psNqoD8A9WgYAPRXzw1B%2B%2BuG4NdMQ4xb%2BDrCgYZGZBr17rcNYeAcgpZfXH6wJmeRsNlH0Hhvo7gyzJC2BKH0/saWPpzOpoAgwwRmbwUSyE/c/N1Gzg%2Bl3JuNcZbtTELVhreTYjh2zZBc%2BVtjN23fi7ZpuFGjdP6cM8Z0z5nDhMes/o9L9mBCOfxdYklxLXMrq82unzG7PHF0C4y4L%2B7QtHv4CeyL4Tot8piYKuJSWkmjwlRPZ9E6wfvtgjlr9%2BWf2/L/fvCrx8QP1JALVjCWHTXtOfkN1DbW7U2vQxNl1EDgEkZ69hsjQbMPjNG/MgXayqM0ciXRyRjGCHOiWxQkzMt1tcc25chY1zs2CbuXm0T0gfQHlVtIBWZbpCfKEbISJojFONqFo9sFLDJYJHiH6M42pjyluLdId5PItbXzu0p73zbdtc2vg9hPfuDgWJNdIIAA%3D%3D).

For O3 test, [see](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahUAUgBMAIUtXyKxqiIEh1ZpgDC6egFc2TA3cAGQImbAA5PwAjbFIQAFZyAAd0JWIXJi9ffwMUtOchELDIthi4xIdsJwyRIhZSIiy/AJ57bEcCplr6oiKI6NiE%2BzqGppzWpRHe0P7SwfiASnt0H1JUTi4fNJMAagA3KoB9Ih2zAGYAER2aenQWInObLQBBLdDgfaqz49PLz9RjmZ4lYzkCLo8zC9Ic89ugCJgdqRvPQIP9vic2KElOQ0T82Cx1NjcSd0DQaCoiDiDqh0TsAFRpABe2B2C1OAHYns9IQBOQSkHao0InAi/K5aR47UXnDw7UFnKxS2y2VnQ3lmTlqrQ83mk8nYB7AqWneJXc5XVGYphKIGK6Wm07WHb4wm241g1n0nZaAB08QhUO1vKZ2DdWhN4KN9rNfytNqjEZ2AFodnqKW7owHuUHtSG3TwIxnE%2BbnQT43bEym0wai2Cs%2Br2eCoY2uEt6Nx4vwAlwdOR0NwPMqnUoVmsWZYznxyERtK2lgBrKQADh9WgAbGv2Vp4jyl%2Byl0v4vEeGdDNxpPw2AktORu73%2B1x%2BEoQDeZz3W%2BQ4LAUBgcPhiGQlDUHQjCsBw3BToIwhiBInAyHIwjKGomjvuQ%2BgWIYximEOdiVNUrgQO4YwtOQwQzCUZRDHk6RCMRuSpDRTB9BR8xtB0NRTHREztFUnTdA0zEDOUww9FxIkCeRQkJEsI6rOs3BvLs1I/CWNx3A8CrQopHzUrSJbKYaIJ1ppgYwnCCI%2BEwSL0CixLOliVJfHiZaOTSPzVpSdkMgQzKehqXINgFOY8h5YbFn8lpYmFHoqi65beombJ0t6fr1sFoVGgWMURfZ1r5sWTpxflHpJSl/omdmOrahlioWOFFq5fFdUxYVZZus1Dqlb65VBZVwY%2BaGRrhnWQ31Y10UOlWZLpqNxlBVVPJ5rNprFQ6JZxqtVxTfqhk7FlK0VequYDRNka1WNG1Gh1W2ptNNZXYWh2BgtS2KvtZ0JdlVxFctN01Z9B3zf1zKbZtYqlq6mWVrdO2g09L0nVDI3nV9ENNdD/3XVmOw4xyTbci2bYdl2s59gOOE7LJY6OmcFj8G%2BOgLEsSDYCwOBxBARNcBe5BXvEN53vwD5Pi%2B06zkz5CLjwPI%2BuyPBaEu27smcPIWPL8SbmeXBnCTqHC2L75LF%2ByBoOgbBJAwsRARAGDm5bcTADwFgYTgewEOsABqBDYAA7gA8kkzAQfwIFELEz4QFEpNRKE9QAJ7B%2BQMesKQcd%2B1Eui8YntscMIftMPQCeoTg%2BImJIxcEKQvEEAcz6odg6hVD4YeJ8K7Sk/QBBRKQ8deDgpNEKQBBXrwH43CwwBKF7vsB0Ho8IaI4iSPBUGKCoGik/orRGCYaA4YYXfPpASzoEknR10mfs62x1cEUR3jNIETCYIJcxxK01GdGJn8ZK/lHcexIQ/FGgP3GDffCXQph/0GBMTioCSKTB6NA9%2BMlRzyR4FzTst5SYPh2OoJca4kxrmkDsYAqBUB7QsD6OqEB/wkAFBODB9NxYLikFQs4a54jsmkBYHhKseEWDOOyLWPMrzywFjg7gItXzi0/IgCAP4zYW0YBQKgNslH2xAI7Z25BXbu2wNPf2gduyQQYGHUgEco6oWTvHRONjU7p0zk4bOZtc5EHzoXUmJcsLl17IQKu1Ra6kwbk3Fu8827tlQp3buvcsAbF7IPYewcljj0noY2eJiQ7yBgsvWQq8kIb1QvoYRO9sLWFsAfKIR9OZ9jPhkC%2BV8nw8QgW4Z%2BYkyLFCkh/BiX94H0XyL/SSb8DB4T4nA7ICDmljKQUM/%2B4kQETJGVA2ZMDUFyU4BYTBut7zcDwQQohJCyEUKdtQwUdCyA002cww2zNWbs2oKwk5KsLD8y0NIfma4tAWE%2BWuERl4pBaAkXrKR9hRYMw/MbSFii7YqOtrbZRgxUClJ4J8m8ejPbeyMXPUx9BzGWOjrHVOdjCVpwzlneeOdmDuILkXPx2BS7AF8fwfx1cgn10bqgZuGwpwRI7l3Huqc%2B7xPpkPEefAUlGDSZijJidV45Lgnk%2BQBSUK9n0KeUpe9ynWEqdUk%2BdShANOvqMjIrTPB9NaB02Ycyf60XNckHpgzOnDIAbfSBok7XGqAcsp1czEGjA9d6q1qzljrKkFs7BwKuB7MIcQnYSKsJ7TXKuVcZzCD0Jpkwg2jNbls0GDUxcLy/RrheWcJc0geTK03DyLQ6rzz/PERGnZj5QUyMNnI%2BACjTYwqtmo%2BFmj40mBRYC3R2A3YYpnsY2VZjw7UCsb2exRcpwLsceSqclK840q8fSnxwrdGV1ZdgOuvYQmcrCTy4Q7con8tif3VCiSxVj0lVPaVk755yqXgqheyrN5DA1eYLVdhom6tqefbgl8jVTJNYRNpdrLUsXfvagZtrFndKQ0xFZCHPVuv9Sh8B0yJI%2BpgfMsSfrpiEZQSGscIBNlaywYLMmUb8ExpIQOj4KLk3hloWmi5E4rlZrnOQFmuaOasMLRrGQ8RnYK0kzyI8fzeYAqBU26R/GjbyOhQi1R1A%2B0qJAHsVASQkiHD2NLQ4ZxvjqGISOsdBiX3YpDtOixs6CUp0XfwZdZLnEUtcVSjxtLmXbrLrullgTD3BI5Vy1uF7Im9miQKuOQqB6iuSQIJ96TX2QWyR%2BqQirELrxVXoajmFd7/psNqoD8A9WgYAPR%2B3UHhqD99cNwa6Yhxi38HWFAwyMyDXr3W4aw8A5BSz%2BuP1gTM8jYbKPoPDfR3BlmSFsCUPp/Y0sfTmdTQBBhgjM3golkJ%2B5%2BbqNnB9LuTca4y3amIWrDW8mxHDrmyC58rbGbtu/F2zTcKNG6f04Z4zpnzOHCY9Z/R6X7MCEc/i6xJLiWuZXV5tdPmN2eOLoFxlwX92haPfwE9kXwnRb5TEwVcSktJNHhKiez6J3g/fbBHLX78s/t%2BX%2B/eFXj4gfqSAWr9WsOmvac/YbqH2t2ptehybLqIHAJI717DZGg2YfGWN%2BZgu1lUZo5EujkjGMEOdMtihJmZYba41ty5CxrnZsE3cvNonpA%2BgPKraQCsy3SE%2BUI2QkTRGKcbULJ7YKWGSwSPEP0ZxtTHlLcW6Q7yeRazpgphtj3m17a5rHhP5uBMHAsSa6QQA%3D).

On the following screenshots, for each compiler, left is rolled, right is unrolled.

O2:

[![loop unrolling](https://dl.illwieckz.net/b/unvanquished/wip/loop-unrolling/20230402-124036-000.godbolt-o2.png)](https://dl.illwieckz.net/b/unvanquished/wip/loop-unrolling/20230402-124036-000.godbolt-o2.png)

O3:

[![loop unrolling](https://dl.illwieckz.net/b/unvanquished/wip/loop-unrolling/20230402-124032-000.godbolt-o3.png)](https://dl.illwieckz.net/b/unvanquished/wip/loop-unrolling/20230402-124032-000.godbolt-o3.png)

Fun stuff: I can force gcc to produce with O3 a similar optimization by asking it to recompute one thing for nothing but this adds that useless compute on any compiler, this is so weird… [see](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahUAUgBMAIUtXyKxqiIEh1ZpgDC6egFc2TEAA2AGZydwAZAiZsADk/ACNsUhAADnIAB3QlYhcmL19/INDM7OchSOi4tkTktIdsJ1yRIhZSIny/AJD7bEcypmbWogrYhKTU%2Bxa2jsLupSnhqNHq8ZSASnt0H1JUTi4fbJMAagA3BoB9IiOzYIARI5p6dBYiG5stAEEDqOBThuDLtc7n9UJczABWKzBCG3N5mT7wj4ndAETBHUjeegQEEAq5sKJKcg4wFsFjqQnEq7oGg0FREIlnVC4o4AKmyAC9sEc1tcAOzvD7wgCcglIR2xUSuBCB9y0byO0puHiO0OCVgVtls3MRwrM/J1WiFwuptOwr0hCuu4PuN3u2PxTCUEPViut12sR1J5Odlph3NZRy0ADpwXDPkcI5GIya6T6tFabcD7QS4wn3eqvU6LfG/TyWYGQ/KAPRFo5IFioADWBqNho52FTMJ9rsT9wdWZdaYAtEcY2bmwmw4LDcL6z6eIOLS2ZZ6yR3fW6e33zZ2m2qa3rYQjebcuBt6NxwfwAlwdOR0NwPJqPUotjsuZZgnxyERtHuNpWpCkg1pAoFeVo4JCikvIpCk4LgjwoQHlw0j8GwIDglo5AnmeF5cPwSggMhr6nnu5BwLAKAYDg%2BDEGQlDUHQjCsBw3DPoIwhiBInAyHIwjKGomh4eQ%2BgWIYximNedj1I0rgQO4MwBDwYRMJgIxVDUiEZFkORCFJBglGpTAKWMyTgj0fRNAsGkyaJ/SDG0ukrPpkxDKZdlWUsinjOCGy3tsuzcN8xyMoCtoPE8LxDj5vyMsyAV%2BSu0LWkOiLIqiRw%2BEwGL0FilKegSDL/CSc7ZUygLLvlzJsgQnL%2BnqAq6vqCIjoay6Nm6AXJo6jU2h6mZtf6%2BbBqG661bWQoNRaE5%2Bs1mWtSNaZap1U25gGvVxXVQ00rGFoWNNSYTfOG1jR1c4%2Brtbp5gWfVVQNo5lQ22aTuqOZNcC7ZdUuq39jda7nYNY7vbFc0PW2KZ/fcL2miuRyjbF/XDl9V1tYdm0A5N6pHcDvavWDKNLdDl2cuOt2Bgjs7ej9qPDXdg5Q7qdaw0DeNjY9B1A0cINreqEOwpTy3fWz%2BMozOs3I92aOg/DH2IlGm46ju%2BEwUeKFvuel7CUcHn3u6wQWPwuE6GsGxINgLA4MkED7twcHkAhSHyzx6GYdhL5vrr5CfjwQpBryPBaCkgG8sEQoWJ74L/oY3DBMeCu2w7eEbIRyBoOgbDpAwSSURAGCJ8nyTADwFj8TgJwELsABqBDYAA7gA8ukzD0fw1FEEkWEQPECvxFErQAJ61%2BQbesKQHcV/EugNLhz7pxwwgV0w9BdzxOCkiYkhzwQpAjwQZxYTx2DqA0PgN93kq9Ar9AEPEpCd14OAK0QpAEAhvD4Y8LDAEoJfl1XNcP%2BxojiJIbGMYoFQGgFb6BkkYEwaBhKGFPlhSAGx0DpH6JvLsFcw6GTXuJSS3hOgGAiM5PSmlVL9Aclpfo1klJmV6BggYJlsGFEoUZIQllFiVAIWZWhBRpKORYcsCh7k7xeR4KbLgctUL8HQkcdQKRAhdkCNII4wBUCoHBhYIMG0IBkRIGKR8QitaOw/FIVRwRAjgl5NICw5i/bmIsMEXkIdYLwSkFoZCYjFYYXsPbbW%2BFY4QGIgnJOjAKBUDTv4zOIBs653IPnQu2A36V2rieBiDAG6kCbi3HivdO7d0yf3Qew8nDd3HswIgU8Z4K3noJJeZ5CCr0aBvBW29d77y/ofGCZ4T5nwvlgPYZ4b531rhsJ%2BL84kf0SXXeQzE/6yAAZxYBPF9B2PAUJawthoHxFgSbc8iDcjINQZhKhYkAgSTkg5PBrCbKENKLkEhRDcjkPGAw6hzCHLmWMkMe5yR2H2ToVw%2BY7z8EXKEZsTynALDCNERHbgkjpGyPkYo5ROc1Hik0WQdWYK9HRz1gbI21ADGIr9hYJCWhpBIUCFoCwZLAj2PNghT2LjIXuKwjhR2BFEA%2BL8RnQJqd04BPGKgJZPAyXIWicXUu8TP5JPoCktJrd2792yXKgeQ8R6FIThPEp09Z7VOwAvYAVT%2BA1LXvUreO9UB7z2M%2BVpx9T7n37pfHpWtb73z4IMowwyxWjO7gAyZrFpnyFmdxM8%2BhoKCUgSs6wayNnwO2UIXZaDXmYJOT83BckPmXO0jcq55QAUUPQYc55ybHn5oWGmr50xC3cNLfwkFUhwXhxtlCqRMi5FHH5YJcGgQfw/mRYQLR6tdFRx1liw24xNmfkJSGQIhLggpGkEKX2/4hRaGgmbRxdLrZoW4HbZl0dWXwF8fHTlKdgk8rCW2kwgrnFROwAXUV78EleuSY3ag6Szw5Nns%2Bd9eSVVfyKZPTV5SdWVIdVEleRrsCbzPI0s1zTLXCCPjxDptqO72uvk6gZAg3Wvw9Q%2Br%2B3rf6%2Bu/gGkBykllhpsBGjpUatlIO4Cg%2BNBz%2BhuCTZwlN8kc0PJUlmvIFbSF3I458vNFkOE4KLcJ/55zc1/PLaxstPCXKfOrfeEAYL7EQobVwaFzb5Hnt%2BIKrt8YNG9tRY%2BdFg73zkH1iO42BiJ1BxkOCXOXtHNCggtStdV7XGRyZeZmObKOW8qCdQU9gSQAnFQOkdI5wTiu3OMEAE6g5HXtvbEnDEq65PtSS%2B2VfcP38C/cqgpv61XFNKVqg1QHF4gcNXUiDDTTXmoPvBtp/AkNdKvjxPpzrH5YZGbhhiEyCNSD9RxIBga9AqYEhA8w4a7BUfgNG2jRYK7qCE7kZjngK1nN4Zxvj6leO3OzZJzjCaaHfNk2tphJaBMGGk%2B0Ctd2q3AuU0C2W9bN2acS/ItgShwunFdkGeLPbyLaJsQOrxTsrM4rHSp4IQZgL/kCLOw0ciA5B3cxbJx9KNPbt83uoih7AvctCaF8LkXouxfi%2BcJtyWYl9fSwITLMqMmKoVbl79RWx4lf/WUuelW9XVbA7VyD/BoONZac161nS7XdLQ/0h%2Brrn7YfvQz/DLFhtEbGyRqlZGZsUbmzAhbNGdkgGW6t07G3TmppuzJPbPGLt29LZds7MnRPO%2BYU7u7LzrvHcU89ryqm3sbvEY26RnpfvKJi27IHRmQdorWBiodlnsWjts9IIMYF/bSC9rO6QZLbGyBgjSrHwe3G44hwYiCIZgiGkgjOqd0gSVCnsZrTH66vNbt88I1vHf3EV/IGcVJ63pBAA%3D). It really means how much more predictable gcc is with unrolled loops…

Reply:
In fact I was right for the matrix compose. In my previous link gcc was using O3, which was not wanted, especially since `RelWithDebInfo` that we uses for release builds use O2.

So, for matrix tranposition.

- gcc with O2 and msvc with O2 and OX only unroll the inner loop, keeping the outer loop, and then adding extra comparison and jump instructions that are totally useless since there is no conditional break or continue.
- clang with O2 and O3 unrolls everything like gcc O3.

By unrolling it by hand we make sure this matrix transposition function gets O3-like optimization in all cases. And yet again, the unrolling is more predictable: we know what we will get, [see](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIM6SuADJ4DJgAcj4ARpjEIADM8aQADqgKhE4MHt6%2B/ilpGQIhYZEsMXGJtpj2jgJCBEzEBNk%2BfgF2mA6Z9Y0ExRHRsQlJCg1NLbntY32hA2VDiQCUtqhexMjsHF7pRgDUAG6dAPoEuybxACK7VLSoTATnVhoAgtuhwAed8Sdnl5/IJxMAFYLPFgRdHiYXlDnvtUHh0LsACrEQwKVJKACy92IeFUEF2aAYo3%2B31OLBxeOBFl28TOQIupFJPwIqOJGMw1Np9KuixhJgA7E9nlCAJyhU54JnaSHQuWi/jEXYEvC/K4aR67VXnNy0zV4azWXZ8%2BWC4VisWK5W7bRq3Ya%2BI0206vWOm2GyzG/kaUUWoXe30%2BsWstEcrnahlc50Mu0U1lUkE2nnh5OOgN%2BiGmgWZkXZ72vHYfQ4A07nK43O4PNMvN57Ytku3FwEgsGR6sil5whHItnotKYbHx1RmAlEkn1n5x3GqLl08FMienEPs/uznle6H%2B%2BXLvtKLkaZOJg/g2OUmdHw8049tp6Bjuindhi/grlcddl3ZThM0t8v583gNH1Xf8IUTMx3z%2BL9zxpcC/yvVNzW3Xsnx/S97Qgq4oP3NDfwApDQ2A1C4N2XCrg/LDE1I18EMA5DCJItDYJjcizy5JjQKIvCO3vIC9zAtDrzIyDWJAtiaPwlc%2BJgnCMM/ETOI43Z2NlHi6KkpTGNkijpOI5T20FC4OGWWhOCBXg/A4LRSFQTg3A9GkFFWdZMDOMx4h4UgCE0IzlgAa38SQADoADYgVFDR4g0LgBS4CKuGCrh9E4SReBYCQNA0UgLKsmyOF4BQQEyrzLKM0g4FgGAoEq8rEBQVAWGSOhYnISg0Aapq4mALgzACLB9jwDYADU8EwAB3AB5ZJGE4DyaFoAhYgKiAom80golCRoAE8Zt4dbmGITbxqibROmKjy2rYQRxoYWhtpK0gsApIxxHu/BiFOvBDgK%2B7MFUTovAWnbyEEapVtoPAolRA6PCwVb4zS7hSpuJhgAUYaxsm6bEZkQQRDEdgpBx%2BQlDUVbdESgwjBQez9AhgrIGWVBklqYlOAAWnGsx8uqD7nAgVwJj8RLglmUpyj0DFCiyTxWglgoWf6MWhkSjoujqaZBb0VWWZ6JpFcGOIVY1mXciN3p9fmQ3lkctYNgkYzTPM1bct2VQAA5grZ4LJF2YBkGQEizEC8CIFwQgSFc9zFl4YqtEWPyQCBTKTI4FLSDSpOsudzh8sKzzvOWGqICQNrGvoMgKAgUuOpALqeoezB%2BqGkaJqmizZroBbiCWlb7r2ragf7g6jpOhwgYuxgCGu27VsewxgBeqy3o%2Br7Vt%2B/7AexiVQfu8HIa2mHNis%2BGduWZHUfR1usdm2Q8fEQn%2BEERQVHUe7dCBfR5%2BpyxrFpqJ6YgIzZmmRvoc3iNzGomQXAMHcCbIWAQRYlANiASQmVJYs01t1fI6QFai2QagqokD1a9Ewe0HmasGC6xmEgy2KDMqjBIXAiQUxzZ4Noag62Tk7ZcAdhwMyWd7ou3dp7b2vt/aB2DsqMORAlTmCjjHAuywECYCYFgOIgDSD%2BUzinNO2VeC5VzkVRRSUOBc3TonTKejrI53ziVQu1Vqol3qmXZqldq7lxAANZAZgNBmGiolPqA1MCX0xu3Xgc0u491WkPO6HkYkj1OuPeql0p43TukvTAT0F5H14MvLoq8fp/WQADTYHlt4pysnvKGm1D5w1xAjHgZ8DAXxbqEoGj9hCiHvtIDpz9SZvxQZ/Kmpgf6WD/gAoBLNQGcwgbzPw/MYGYMCDAi24tEroMyEsjZRQ2FrMIXMqhSztbdGmKs5WthjY5CFhc1hNC1mcNtgTXh/CrFCI9l7H2XilK%2BK4IFAUgU3yh3wDIyOPCFF2KUSotRlAE7aOSqlCxAico2IKkYuxZVHF1XauXFqVdnE1y8T4vxMUG5N2Ca0tu7TO6LUoL3KyMTB4bWHsdRJ2MJ5XTSbPTJ89F65LwO9fJmBvpWXXsUzeZSQYVN4FUg%2BGAcmeXqafPgzS0YUuvuE2%2BXSCY9NkH01%2BVldDBSGcYGme8JnWWAQIaZ4D9kUOgbAq5KCEErN2QsbBUtNaSACNshgZyKi2p1pc2WXqA0nNuXMcWIwg25BDQwvWrqKgPOcvbExLzs4cFdu80RXyiW/P%2BYC6REc5FgtsXHSFqihgaK0cneFSL9EorzrHHyJizEZ0sem8FZbNEoP%2BaKLgoogRAgSjFUUbtEhGpTja15Nim32PgEXTF7jXGtXxR45AlMurBQyqSoJITKXYwiTS5a0SmWxN2qehJY82XJMntPdJuTuXPXlXkxwBSRVFJKUDcpYMIbVNqfdE%2BiMmko1VRjfdN9cZaokDqp%2BJN9U6ASMa7%2BVgxlmvgJMkB7MZmhr5gLJhwsXV3POT6rZ8tMh%2Bq1uQwNjDHUqyo2G%2BNRHDY3PGPhlj1CI3KyTdw55TtBGcEzSIn26754kWCoFDQEmpHAqLW5Ets7y3QqrYinRCLM7TryrYRtBcMXzqcdi5deKDNxBE0YeK27AnNzA%2Bqvg1Lu60pPftM9a0L0sqvedG9HKZ6vUfdk2e/KV5CrXh%2B8VvBv271/bK2GAHFVAeVSBvdNmOl321UTPVZMQBuyQyMlDNg0MMwtVMrDNrjm4cWWxxBnHmMkbYz6ijdGiGUOjdc0rTXw1K2Y3G5obGuv1e4081NfHkUZuER8wkG6xMSak0C8Osi5PR1Lc25RFb1GwpranIb9bNOosW/HFtan238c07O3TFU9NYpcRXFdxmQD7GQMkZIRx9j9qOIkI4qhvY7qs1fMJtn5pHrpeepzjKnOXrOrwdlqTvMZKybyh6AXBXCt4KKz9W9JU/v3tDOVdS8ANKRiqxLv3ktQYfrquDGWAgbuQ7/fLGimZFYAPScxw/MvDtHlnoHq%2B6jBtWyM7KY5Rxrhyev0eIYxqrWtmuS/a8gnhKxHkppTmmo7rtREsAUHdg4/bAqJGk7N0FC2FOkGW0ptbJi05trrdY7b2n0VF305d3FS64h3Ye09l7b2PvSEs%2BS6zRO7NRL7qekHW0wdJJYCku9XLYfPoR6%2BoLhSN6lLC%2BjiLmOanY5i7jpV59QM/faZq/G0G0vk4GQKbLpq6bocK5hpnqgWf2qWZVjrcscGbN523/nEuGsHKlz3ihVCuddaOachNKabbJrl0rzb1vVc%2B3V5r57oodd0hmyC4thvjEm8rWb1TVuDFabRV2lOrbEUac7c2rRQcNBuw0JISQ8Q3ZmFv/FT7k6Z8H6N/5AUQIdeoNFMFKOs/gAWOllifgirFIdsNhfntuAfvjOsYocN3FApIEAA%3D%3D).

[![loop unrolling](https://dl.illwieckz.net/b/unvanquished/wip/loop-unrolling/20230402-153317-000.godbolt-matrix-transpose.png)](https://dl.illwieckz.net/b/unvanquished/wip/loop-unrolling/20230402-153317-000.godbolt-matrix-transpose.png)

Reply:
So is there anything left that would block that unrolling? Such unrolling produces more predictable code, and actually a good one, so there is more chance to get unoptimal code and different code among compilers when not unrolling. Also it produces good code on both gcc and clang, which are the compilers used for release, and in fact it produces better code than what gcc does the way we use it to build release builds.